File: img_bmp/aritmetica/aritmetica_hilos_buffer/aritmetica.java
import java.io.FileOutputStream;
import java.io.IOException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class aritmetica {    

    private static final int N_THREADS = 4;
    private static int n_padding_byte;
    private static byte[] data;

    // estructura equivalente a la estructura threadArgs en C/C++
    private static class ThreadArgs {
        int width;
        int height;
        int start_height;
        int end_height;
    }

    // estructura que llama a la funcion run de la clase Thread
    private static class MyThread extends Thread {
        private ThreadArgs args;

        public MyThread(ThreadArgs args) {
            this.args = args;
        }

        public void run() {
            // tamanio teniendo en cuenta args.end_height - args.start_height
            int index = args.start_height * (args.width * 3 + n_padding_byte);

            // asigna los colores
            for (int y = args.start_height; y < args.end_height; y++) {
                for (int x = 0; x < args.width; x++) {
                    data[index + 2] = (byte) ((x / (float) args.width) * 255);
                    data[index + 1] = (byte) (((args.height - y) / (float) args.height) * 255);
                    data[index] = (byte) ((y / (float) args.height) * 255);

                    index += 3;
                }
                index += n_padding_byte;
            }

        }
    }

    private static void createBMP(int width, int height) {
        
        // n_padding_byte = (4 - (width * 3) % 4) % 4;
        n_padding_byte = 4 - ((width * 3) % 4);
        if (n_padding_byte == 4) {
            n_padding_byte = 0;
        }

        // cabecera del archivo BMP
        int size = 54 + ((width * 3) + n_padding_byte) * height;
        int image_size = ((width * 3) + n_padding_byte) * height;

        // alineaminto usando un buffer
        // java por defecto usa big endian
        ByteBuffer buffer = ByteBuffer.allocate(54).order(ByteOrder.LITTLE_ENDIAN);
        buffer.putShort((short) 0x4D42); // type
        buffer.putInt(size);             // size
        buffer.putShort((short) 0);      // reserved1
        buffer.putShort((short) 0);      // reserved2
        buffer.putInt(54);                // offset
        buffer.putInt(40);                // header_size
        buffer.putInt(width);             // width
        buffer.putInt(height);            // height
        buffer.putShort((short) 1);       // planes
        buffer.putShort((short) 24);      // bitsperpixel
        buffer.putInt(0);                 // compression
        buffer.putInt(image_size);        // image_size
        buffer.putInt(0);                 // x_resolution
        buffer.putInt(0);                 // y_resolution
        buffer.putInt(0);                 // used_colors
        buffer.putInt(0);                 // important_colors

        //apertura del archivo
        try (FileOutputStream file = new FileOutputStream("img_java.bmp")) {
            // escribe la cabecera
            file.write(buffer.array());
            
            // buffer para la imagen
            int data_size = (width * 3 + n_padding_byte) * height;
            data = new byte[data_size];

            // parte hilo
            MyThread[] threads = new MyThread[N_THREADS];
            int chunk_size = height / N_THREADS;
            ThreadArgs[] t_args = new ThreadArgs[N_THREADS];

            for (int i = 0; i < N_THREADS; i++) {
                t_args[i] = new ThreadArgs();
                t_args[i].width = width;
                t_args[i].height = height;
                t_args[i].start_height = i * chunk_size;
                t_args[i].end_height = (i < N_THREADS - 1) ? t_args[i].start_height + chunk_size : height;
                threads[i] = new MyThread(t_args[i]);
                threads[i].start();
            }

            for (int i = 0; i < N_THREADS; i++) {
                try {
                    threads[i].join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }

            file.write(data);
            
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) {
        int width = 8000;
        int height = 8000;

        createBMP(width, height);
    }
}

File: img_bmp/aritmetica/aritmetica_hilos_buffer/aritmetica.c
#include <stdio.h>
#include <stdint.h>
#include <pthread.h>
#include <stdlib.h>
#include <string.h>

#define N_THREADS 4

// alineamiento de la memoria
#pragma pack(push, 1)

typedef struct {
    uint16_t type;
    uint32_t size;
    uint16_t reserved1;
    uint16_t reserved2;
    uint32_t offset;
    uint32_t header_size;
    int32_t width;
    int32_t height;
    uint16_t planes;
    uint16_t bitsperpixel;
    uint32_t compression;
    uint32_t image_size;
    int32_t x_resolution;
    int32_t y_resolution;
    uint32_t used_colors;
    uint32_t important_colors;
} BMPHeader;

//alineamiento normal
#pragma pack(pop)

// variables globales
uint32_t n_padding_byte;
char *data;

// argumentos para el hilo
typedef struct {
    int width;
    int height;
    int start_height;
    int end_height;
} threadArgs;

void *generateColor(void *t_args){
    threadArgs args;
    memcpy (&args, t_args, sizeof(threadArgs));

    // colores de la imagen
    int index = args.start_height * (args.width * 3 + n_padding_byte);
    for (int y = args.start_height; y < args.end_height; y++) {
        for (int x = 0; x < args.width; x++) {
            data[index + 2] = (char)((x / (float)args.width) * 255);
            data[index + 1] = (char)(((args.height - y) / (float)args.height) * 255);
            data[index] = (char)((y / (float)args.height) * 255);
            
            index += 3;
        }
        index += n_padding_byte;
    }

    pthread_exit(NULL);
}


void createBMP(int width, int height) {
    
    n_padding_byte = 4 - ((width * 3) % 4);
    if (n_padding_byte == 4){
        n_padding_byte = 0;
    }

    // formato de cabeceras
    BMPHeader header = {
        .type = 0x4D42, 
        .size = sizeof(BMPHeader) + ((width * 3) + n_padding_byte) * height,
        .reserved1 = 0,
        .reserved2 = 0,
        .offset = sizeof(BMPHeader),
        .header_size = 40,
        .width = width,
        .height = height,
        .planes = 1,
        .bitsperpixel = 24,
        .compression = 0,
        .image_size = ((width * 3) + n_padding_byte) * height,
        .x_resolution = 0,
        .y_resolution = 0,
        .used_colors = 0,
        .important_colors = 0
    };

    // apertura del archivo
    FILE* file = fopen("img_c.bmp", "wb");
    if (!file) {
        printf("Error al abrir el archivo.\n");
        return;
    }

    // escritura de la cabecera en el archivo
    fwrite(&header, sizeof(BMPHeader), 1, file);

    // buffer para la imagen
    size_t data_size = (width * 3 + n_padding_byte) * height;
    data = (char*)calloc(data_size, sizeof(char));

    // parte hilo
    pthread_t threads[N_THREADS];
    int chunk_size = height / N_THREADS;
    threadArgs t_args[N_THREADS];
    
    for (int i = 0; i < N_THREADS; i++) {
        t_args[i].width = width;
        t_args[i].height = height;
        t_args[i].start_height = i * chunk_size;
        t_args[i].end_height = (i < N_THREADS - 1) ? t_args[i].start_height + chunk_size : height;
        pthread_create(&threads[i], NULL, generateColor, &t_args[i]);
    }

    for (int i = 0; i < N_THREADS; i++) {
        pthread_join(threads[i], NULL);
    }

    fwrite(data, sizeof(char), data_size, file);
    free(data);
    fclose(file);
}

int main() {
    int width = 8000;
    int height = 8000;

    createBMP(width, height);
    return 0;
}